AcademyApp/ViewControllers/LoginViewController.swift

```diff
@@ -0,0 +1,189 @@
+    @IBOutlet weak var passwordTextField: UITextField!
+    @IBOutlet weak var doneButton: UIButtonX!
+    
+    @IBOutlet weak var showButton: UIButton!
+    
+    @IBAction func backgroundTapped(_ sender: UITapGestureRecognizer) {
+        self.view.endEditing(true)
+    }
+    
+    @IBAction func emailTextFieldEditingDidChange(_ sender: UITextField) {
+        doneButton.isEnabled = isValidEmail(possibleEmail: sender.text!) && passwordTextField.text!.count > 0 ? true : false
+    }
+    
+    @IBAction func passwordTextFieldEditingDidChange(_ sender: UITextField) {
+        UIView.animate(withDuration: 0.2) {
+            self.showButton.isHidden = sender.text!.count > 0 ? false : true
+        }
+        
+        doneButton.isEnabled = isValidEmail(possibleEmail: emailTextField.text!) && passwordTextField.text!.count > 0 ? true : false
+    }
+    
+    // control event : touch down (hold)
+    @IBAction func showPassword(_ sender: UIButton) {
+        passwordTextField.isSecureTextEntry = false
+    }
+    
+    // control event : touch up inside (release)
+    @IBAction func hidePassword(_ sender: UIButton) {
+        passwordTextField.isSecureTextEntry = true
+    }
+}
+
+// MARK - UITextField's delegate methods
+extension LoginViewController : UITextFieldDelegate
+{
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+        self.view.endEditing(true)
+        return true
+    }
+}
+
+// MARK: - Responding to keyboard notifications
+extension LoginViewController
+{
+    private func addObservers(){
+        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillShow(notification:)), name: UIResponder.keyboardWillShowNotification, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillHide(notification:)), name: UIResponder.keyboardWillHideNotification, object: nil)
+    }
+    
+    private func removeObservers(){
+        NotificationCenter.default.removeObserver(self, name: UIResponder.keyboardWillShowNotification, object: nil)
+        NotificationCenter.default.removeObserver(self, name: UIResponder.keyboardWillHideNotification, object: nil)
+    }
+    
+    @objc func keyboardWillShow(notification: Notification){
+            
+        guard let keyboardRect = (notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? NSValue)?.cgRectValue  else {
+            return
+        }
+            
+        let keyboardHeight = keyboardRect.height
+        doneButtonBottomConstraint.constant =  keyboardHeight
+        UIView.animate(withDuration: 0.5) {
+            self.view.layoutIfNeeded()
+        }
+    }
+    
+    @objc func keyboardWillHide(notification: Notification){
+        
+        doneButtonBottomConstraint.constant =  defaultDoneButtonBottomSpacing
+        UIView.animate(withDuration: 0.5) {
+            self.view.layoutIfNeeded()
+        }
+    }
+}
+
+// MARK: - UI Setup Methods
+extension LoginViewController
```

Comment:
For extension with private methods I would suggest private whole extension

Reply:
Updated extensions to be private if needed.